gateway/request_context.go

```diff
@@ -281,32 +282,102 @@ func (rc *requestContext) BuildProtocolContextFromHTTP(httpReq *http.Request) er
 		return err
 	}
 
-	// TODO_TECHDEBT(@commoddity): implement a service-specific retry mechanism based on the protocol's response/error:
-	// This would need to distinguish between:
-	// 1) protocol errors, e.g. when an endpoint is maxed out for a service+app combination,
-	// 2) QoS errors, e.g.:
-	// 	A. The request is invalid: e.g. a JSONRPC request with no specified method.
-	//	B. An endpoint returns an invalid response.
-	//
-	// TODO_FUTURE: Support multiple concurrent relays to multiple endpoints for a single user request.
-	// e.g. for handling JSONRPC batch requests.
-	rc.qosCtx.UpdateWithResponse(endpointResponse.EndpointAddr, endpointResponse.Bytes)
+	// Recursively retry with the new endpoint
+	return rc.handleRelayRequestWithRetries(maxRetries, currentAttempt+1, triedEndpoints)
+}
+
+// selectNewEndpointForRetry attempts to select a new endpoint and rebuild the protocol context.
+func (rc *requestContext) selectNewEndpointForRetry(
+	triedEndpoints []protocol.EndpointAddr,
+) error {
+	// Get available endpoints for the service
+	availableEndpoints, endpointLookupObs, err := rc.protocol.AvailableEndpoints(rc.context, rc.serviceID, nil)
+	if err != nil {
+		rc.updateProtocolObservations(&endpointLookupObs)
+		return fmt.Errorf("error getting available endpoints for retry: %w", err)
+	}
+
+	// Filter out already tried endpoints using helper
+	filteredEndpoints := filterUntriedEndpoints(availableEndpoints, triedEndpoints)
+
+	// If no new endpoints are available, fall back to the QoS selector with all endpoints
+	var selectedEndpointAddr protocol.EndpointAddr
+	if len(filteredEndpoints) == 0 {
+		rc.logger.Warn().Msg("No new endpoints available for retry, using QoS selector with all endpoints.")
+		selectedEndpointAddr, err = rc.qosCtx.GetEndpointSelector().Select(availableEndpoints)
+	} else {
+		rc.logger.Info().Msgf("Selecting new endpoint out of %d available endpoints for retry.", len(filteredEndpoints))
+		selectedEndpointAddr, err = rc.qosCtx.GetEndpointSelector().Select(filteredEndpoints)
+	}
+	if err != nil {
+		rc.updateProtocolObservations(&endpointLookupObs)
+		return fmt.Errorf("error selecting new endpoint for retry: %w", err)
+	}
+
+	// Build a new protocol context for the selected endpoint
+	protocolCtx, protocolCtxSetupErrObs, err := rc.protocol.BuildRequestContextForEndpoint(rc.context, rc.serviceID, selectedEndpointAddr, rc.httpReq)
+	if err != nil {
+		rc.updateProtocolObservations(&protocolCtxSetupErrObs)
+		return fmt.Errorf("error building protocol context for retry: %w", err)
+	}
+
+	// Update the protocol context
+	rc.protocolCtx = protocolCtx
+	rc.logger.Info().Msgf("Successfully selected new endpoint %s for retry.", selectedEndpointAddr)
 
 	return nil
 }
 
+// filterUntriedEndpoints returns a list of endpoints from availableEndpoints that are not present in triedEndpoints.
+func filterUntriedEndpoints(availableEndpoints, triedEndpoints protocol.EndpointAddrList) protocol.EndpointAddrList {
```

Comment:
Same comment re redundancy: any endpoint that fails to return a valid response is already filtered.
If the sanctioning is not working as expected in some cases, that is where analysis and/or code change should happen IMO.